angr/engines/pcode/lifter.py

```diff
@@ -0,0 +1,1506 @@
+#     - Eliminate Vex references where possible
+#     - Consider moving pieces of lifter classes to higher abstraction layer
+#       to reduce duplication with Vex
+#     - Fix default_exit_target
+#     - Fix/remove NotImplementedError's
+
+from collections import defaultdict
+import pypcode
+import cle
+from archinfo import ArchARM
+from cachetools import LRUCache
+import logging
+import os
+import sys
+
+from.behavior import BehaviorFactory
+from ..engine import SimEngineBase
+from ...state_plugins.inspect import BP_AFTER, BP_BEFORE
+from ...misc.ux import once
+from ...errors import SimEngineError, SimTranslationError, SimError
+from ... import sim_options as o
+
+l = logging.getLogger(__name__)
+
+IRSB_MAX_SIZE = 400
+IRSB_MAX_INST = 99
+MAX_INSTRUCTIONS = 99999
+MAX_BYTES = 5000
+
+
+class PyVEXError(Exception):
+    pass
+
+
+class SkipStatementsError(PyVEXError):
+    pass
+
+
+class LiftingException(Exception):
+    pass
+
+
+class NeedStatementsNotification(LiftingException):
+    """
+    A post-processor may raise a NeedStatementsNotification if it needs to work with statements, but the current IRSB
+    is generated without any statement available (skip_stmts=True). The lifter will re-lift the current block with
+    skip_stmts=False upon catching a NeedStatementsNotification, and re-run the post-processors.
+
+    It's worth noting that if a post-processor always raises this notification for every basic block without statements,
+    it will essentially disable the skipping statement optimization, and it is bad for performance (especially for
+    CFGFast, which heavily relies on this optimization). Post-processor authors are encouraged to at least filter the
+    IRSBs based on available properties (jumpkind, next, etc.). If a post-processor must work with statements for the
+    majority of IRSBs, the author should implement it in PyVEX in C for the sake of a better performance.
+    """
+
+
+class IRSB:
+    """
+    IRSB stands for *Intermediate Representation Super-Block*. An IRSB in is a
+    single-entry, multiple-exit code block.
+
+    :ivar arch:             The architecture this block is lifted under
+    :vartype arch:          :class:`archinfo.Arch`
+    :ivar statements:       The statements in this block
+    :vartype statements:    list of :class:`IRStmt`
+    :ivar next:             The expression for the default exit target of this block
+    :vartype next:          :class:`IRExpr`
+    :ivar int offsIP:       The offset of the instruction pointer in the VEX guest state
+    :ivar int stmts_used:   The number of statements in this IRSB
+    :ivar str jumpkind:     The type of this block's default jump (call, boring, syscall, etc) as a VEX enum string
+    :ivar bool direct_next: Whether this block ends with a direct (not indirect) jump or branch
+    :ivar int size:         The size of this block in bytes
+    :ivar int addr:         The address of this basic block, i.e. the address in the first IMark
+    """
```

Comment:
We might want to move type annotations into the method declaration of `__init__()`.

Reply:
I've followed the template of pyvex lifter here. If you want to move annotations into the method decl we can do that.

Reply:
Yeah we are gradually moving to using type annotations in Python 3.6+.

Reply:
Done